src/analysis/vector/geometry_checker/qgsgeometrycheckerutils.h

```diff
@@ -14,181 +14,195 @@
-  {
-    public:
-
-      /**
-       * Create a new layer/feature combination.
-       * The layer is defined by \a pool, \a feature needs to be from this layer.
-       * If \a useMapCrs is True, geometries will be reprojected to the mapCrs defined
-       * in \a context.
-       */
-      LayerFeature( const QgsFeaturePool *pool, const QgsFeature &feature, QgsGeometryCheckerContext *context, bool useMapCrs );
+  public:
+    class ANALYSIS_EXPORT LayerFeature
+    {
+      public:
+
+        /**
+         * Create a new layer/feature combination.
+         * The layer is defined by \a pool, \a feature needs to be from this layer.
+         * If \a useMapCrs is True, geometries will be reprojected to the mapCrs defined
+         * in \a context.
+         */
+        LayerFeature( const QgsFeaturePool *pool, const QgsFeature &feature, const QgsGeometryCheckContext *context, bool useMapCrs );
+
+        /**
+         * Returns the feature.
+         * The geometry will not be reprojected regardless of useMapCrs.
+         */
+        const QgsFeature &feature() const;
+
+        /**
+         * The layer.
+         */
+        QPointer<QgsVectorLayer> layer() const SIP_SKIP;
+
+        /**
+         * The layer id.
+         */
+        QString layerId() const;
+
+        /**
+         * Returns the geometry of this feature.
+         * If useMapCrs was specified, it will already be reprojected into the
+         * CRS specified in the context specified in the constructor.
+         */
+        const QgsGeometry &geometry() const;
+        QString id() const;
+        bool operator==( const LayerFeature &other ) const;
+        bool operator!=( const LayerFeature &other ) const;
+
+        /**
+         * Returns if the geometry is reprojected to the map CRS or not.
+         */
+        bool useMapCrs() const;
+
+      private:
+        const QgsFeaturePool *mFeaturePool;
+        QgsFeature mFeature;
+        QgsGeometry mGeometry;
+        bool mMapCrs;
+    };
+
+    class ANALYSIS_EXPORT LayerFeatures
+    {
+      public:
+#ifndef SIP_RUN
+        LayerFeatures( const QMap<QString, QgsFeaturePool *> &featurePools,
+                       const QMap<QString, QgsFeatureIds> &featureIds,
+                       const QList<QgsWkbTypes::GeometryType> &geometryTypes,
+                       QgsFeedback *feedback,
+                       const QgsGeometryCheckContext *context,
+                       bool useMapCrs = false );
+
+        LayerFeatures( const QMap<QString, QgsFeaturePool *> &featurePools,
+                       const QList<QString> &layerIds, const QgsRectangle &extent,
+                       const QList<QgsWkbTypes::GeometryType> &geometryTypes,
+                       const QgsGeometryCheckContext *context );
+
+        class iterator
+        {
+          public:
+            iterator( const QList<QString>::const_iterator &layerIt, const LayerFeatures *parent );
+            iterator( const iterator &rh );
```

Comment:
@nyalldawson short summary:

* `mCurrentFeature` has been changed to `std::unique_ptr`.
* Compilation fails because "copy constructor has been deleted" in `QgsGeometryCheckerUtils::LayerFeatures::begin()`.
* Makes sense, it could return a copy there
* But AFAICT it uses copy elision
* So now, there's a (potentialy expensive) copy constructor that is required but not used
* And in case we actually at some point do have a copy of such an iterator we will not get a double delete (which is another hint that the copy constructor was not actually used before)

Any better idea to deal with this appreciated

Reply:
Guess it's not possible -- Qt containers just don't play well with unique_ptr. You could maybe swap the container out for a std::vector, but not sure if that breaks something somewhere else.

Reply:
Which Qt container?
It's the `std::unique_ptr` which deletes the copy constructor and the iterator is for a home brewed container.